tests/blackbox.py

```diff
@@ -440,3 +477,412 @@ def dryrun_one(
+            are of type PredicateKind.IdenticalPair, this parameter needs to be
+            provided for comparison.
+            NOTE: model_router may in fact be the same as router, and in this case
+            it is expected that something else such as version or optimization option
+            would differ between model_router and router during the simulation
+        * algod (optional) - if missing, just get one
+
+    Artifacts from Step 1 are stored in self.results: _SimConfig
+
+    2. Simulation (simulate_and_assert method): - using self.results artifacts Step 1, also takes params:
+        * approval_arg_strat_type: Type[ABICallStrategy] - strategy type to use for approval program's arg generation
+        * clear_arg_strat_type: Type[ABICallStrategy] | None - strategy type to use for clear program's arg generation
+        * approval_abi_args_mod: ABIArgsMod (default None) - used to specify any arg mutation
+        # TODO: currently there aren't any clear_abi_args_mod, but we might need for testing non-trivial clear programs
+        * version: int - for compiling self.router
+        * method_configs: ABICallConfigs - these drive all the test cases
+        * assemble_constants: bool (optional) - for compiling self.router
+        * optimize: OptimizeOptions (optional) - for compiling self.router
+        * num_dryruns: int (default 1) - the number of input runs to generate
+            per method X config combination
+        * txn_params: TxParams (optional) - other TxParams to append
+            -in addition to the (is_app_create, OnComplete) information
+        * model_version: int - for compiling self.model_router
+        * model_assemble_constants: bool (optional) - for compiling self.model_router
+        * model_optimize: OptimizeOptions (optional) - for compiling self.model_router
+        * msg: string (optional) - message to report when an assertion is violated
+    """
+
+    def __init__(
+        self,
+        router: Router,
+        predicates: CallPredicates,
+        *,
+        model_router: Router | None = None,
+        algod: v2client.algod.AlgodClient | None = None,
+    ):
+        self.router: Router = self._validate_router(router)
+
+        self.predicates: CallPredicates = self._validate_predicates(predicates)
+
+        self.model_router: Router | None = None
+        if model_router:
+            self.model_router = self._validate_router(model_router, kind="Model")
+
+        self.algod: v2client.algod.AlgodClient = algod or algod_with_assertion()
+
+        self.auto_path_results: Dict[
+            str | None, dict[tuple[bool, OnComplete], SimulationResults]
+        ] = _AutoPathDict()
+
+    # ---- Validation ---- #
+
+    @classmethod
+    def _validate_router(cls, router: Router, kind: str = "Base") -> Router:
+        assert isinstance(router, Router), (
+            f"Wrong type for {kind} Router: {type(router)}. Please provide: " f"Router."
```

Comment:
what is the `Please provide: " f"Router.`? seems like fat finger?

Reply:
thx for catching

Reply:
see https://github.com/algorand/pyteal/pull/634/commits/5b98699bd363e8290219c57fff41942fb7ca60dc

Reply:
looks good, resolving